src/Microsoft.AspNetCore.ResponseCaching/Internal/DistributedResponseCacheStore.cs

```diff
@@ -4,28 +4,46 @@
 using System;
 using System.Threading.Tasks;
 using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Options;
+using Microsoft.AspNetCore.Builder;
 
 namespace Microsoft.AspNetCore.ResponseCaching.Internal
 {
     public class DistributedResponseCacheStore : IResponseCacheStore
     {
         private readonly IDistributedCache _cache;
+        private readonly ResponseCacheOptions _options;
 
-        public DistributedResponseCacheStore(IDistributedCache cache)
+        public DistributedResponseCacheStore(IDistributedCache cache, IOptions<ResponseCacheOptions> options)
         {
             if (cache == null)
             {
                 throw new ArgumentNullException(nameof(cache));
             }
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
 
             _cache = cache;
+            _options = options.Value; // This store needs its own options
         }
 
-        public async Task<object> GetAsync(string key)
+        public async Task<IResponseCacheEntry> GetAsync(string key)
         {
             try
             {
-                return CacheEntrySerializer.Deserialize(await _cache.GetAsync(key));
+                var entry = ResponseCacheEntrySerializer.Deserialize(await _cache.GetAsync(key));
+
+                var cachedResponse = entry as CachedResponse;
+                if (cachedResponse != null)
+                {
+                    var shardCount = (cachedResponse.Body.Length + _options.CachedBodyShardSize - 1) / _options.CachedBodyShardSize;
+                    var copyOnlyDistributedCacheStream = (CopyOnlyDistributedCacheStream)cachedResponse.Body;
```

Comment:
Curious if that's the style we have everywhere else? I have always skipped the space and I see it elsewhere: https://github.com/aspnet/Common/blob/dev/src/Microsoft.Extensions.Primitives/StringValues.cs#L390-L400